### PR TITLE
Convert time-format to TypeScript

### PR DESCRIPTION
**Pull request by kristw**
🏆 Enhancements

- Convert `@superset-ui/time-format` to TypeScript

💔 Breaking Changes

- BREAKING CHANGE: Make formatter always returns `string`. This is different from previous behavior.

|Value|formatted value (before)|formatted value (after)|
|------------|------------|-----------|
| `null` | `null` | `'null'` |
| `undefined` | `undefined` | `'undefined'` |
| `NaN` | `NaN` | `'NaN'` |

### CODE REVIEW

**Comment by mistercrunch** [comment]
Are we sure there are no truthy evaluation of what comes out of this function in the Superset codebase? Wouldn't it be safer to either return either `null` or a string, or an falsy empty string?

File: packages/superset-ui-time-format/src/utils.ts [hunk 1]
```diff
@@ -29,23 +29,24 @@ import {
   utcSaturday,
   utcMonth,
   utcYear,
+  CountableTimeInterval,
```
Comment (williaster) [inline packages/superset-ui-time-format/src/utils.ts]: nice

File: packages/superset-ui-time-format/test/factories/createD3TimeFormatter.test.js [hunk 2]
```diff
@@ -1,32 +1,8 @@
 import createD3TimeFormatter from '../../src/factories/createD3TimeFormatter';
-import { PREVIEW_TIME } from '../../src/TimeFormatter';
-import { TimeFormats } from '../../src';
 
 describe('createD3TimeFormatter(config)', () => {
   it('requires config.formatString', () => {
```
Comment (williaster) [inline packages/superset-ui-time-format/test/factories/createD3TimeFormatter.test.js]: any reason to not just use `@ts-ignore` and have a single (`.ts`) file?

File: packages/superset-ui-time-format/test/TimeFormatter.test.js [hunk 3]
```diff
@@ -1,9 +1,14 @@
-import TimeFormatter, { PREVIEW_TIME } from '../src/TimeFormatter';
+import TimeFormatter from '../src/TimeFormatter';
 
 describe('TimeFormatter', () => {
```
Comment (williaster) [inline packages/superset-ui-time-format/test/TimeFormatter.test.js]: any reason to not just use `@ts-ignore` and have a single (`.ts`) file?

**Comment by kristw** [comment]
My understanding of what @mistercrunch means is when people do sth like 

```js
formatTime(time, 'formatString') || 'custom label when null or undefined'
```

This will work if the `formatTime()` function returns truthy/falsy. My rationale for making it string was to make the function complete in itself without relying on `|| 'fallback'`. If dev want to manually handle labels for `null` or `undefined` then they can add stricter check before calling `formatTime()`.

File: packages/superset-ui-time-format/test/factories/createD3TimeFormatter.test.js (hunk 2 above)
Comment (kristw) [reply packages/superset-ui-time-format/test/factories/createD3TimeFormatter.test.js]: This was written before I know about `@ts-ignore`. Will update.

**Pull request merged**